File: src/actionlib_basics/src/fibonacci_server.cpp
#include <ros/ros.h>
#include <actionlib/server/simple_action_server.h>
#include <actionlib_basics/FibonacciAction.h>

class FibonacciAction {
protected:
    ros::NodeHandle nh = ros::NodeHandle();

    // NodeHandle instance must be created before this line. Otherwise strange error occurs.
    actionlib::SimpleActionServer<actionlib_basics::FibonacciAction> action_server;

    std::string action_name;

    actionlib_basics::FibonacciFeedback feedback;
    actionlib_basics::FibonacciResult result;
public:
    FibonacciAction(std::string name) :
        action_server(nh, name, boost::bind(&FibonacciAction::executeCallback, this, _1), false),
        action_name(name) {
            action_server.start();
        }
    
    ~FibonacciAction(void) {}

    void executeCallback(const actionlib_basics::FibonacciGoalConstPtr &goal) {
        // helper variables
        auto rate = ros::Rate(1);
        bool success = true;

        // push back the seeds for the Fibonacci sequence
        feedback.sequence.clear();
        feedback.sequence.push_back(0);
        feedback.sequence.push_back(1);

        // publish info to the console for the user
        ROS_INFO("%s: Executing, creating Fibonacci sequence of order %li with seeds %i, %i", action_name.c_str(), goal->order, feedback.sequence[0], feedback.sequence[1]);
    
        // start executing the action
        for (int i = 1; i <= goal->order; i++) {
            if (action_server.isPreemptRequested() || !ros::ok()) {
                ROS_INFO("%s: Preempted", action_name.c_str());
                action_server.setPreempted();
                success = false;
                break;
            }

            feedback.sequence.push_back(feedback.sequence[i] +feedback.sequence[i - 1]);

            action_server.publishFeedback(feedback);

            rate.sleep();
        }

        if (success) {
            result.sequence = feedback.sequence;
            ROS_INFO("%s: Succeeded", action_name.c_str());
            action_server.setSucceeded(result);
        }
    }
};

int main(int argc, char **argv) {
    ros::init(argc, argv, "fibonacci");

    // Previously, I'm creating a FibonacciAction object with auto:
    // ```
    // auto fibonacci = FibonacciAction("fibonacci");
    // ```
    // which tries to make a copy.
    // But FibonacciAction contains a SimpleActionServer member,
    // which internally has boost::recursive_mutex, boost::condition_variable_any, and boost::mutex,
    // all of which are non-copyable (deleted copy constructors).
    // SimpleActionServer is move-only or not copyable-by-default
    // because it manages thread synchronization primitives that shouldn't be copied.
    // To avoid such issue, use direct-initialization. 
    FibonacciAction fibonacci("fibonacci");
    
    ros::spin();
}

File: src/actionlib_basics/scripts/fibonacci_server.py
#!/usr/bin/env python3

import rospy
import actionlib
import actionlib_basics.msg


class FibonacciAction(object):
    def __init__(self, name: str):
        self.feedback = actionlib_basics.msg.FibonacciFeedback()
        self.result = actionlib_basics.msg.FibonacciResult()

        self.action_name = name
        self.action_server = actionlib.SimpleActionServer(
            self.action_name,
            actionlib_basics.msg.FibonacciAction,
            execute_cb=self.execute_callback,
            auto_start=False,
        )

        self.action_server.start()

    def execute_callback(self, goal: actionlib_basics.msg.FibonacciGoal) -> None:
        rate = rospy.Rate(1)
        success = True

        # append the seeds for Fibonacci sequence
        self.feedback.sequence.clear()
        self.feedback.sequence.append(0)
        self.feedback.sequence.append(1)

        rospy.loginfo(f"{self.action_name}: Executing, creating Fibonacci sequence of order {goal.order} with seeds {self.feedback.sequence[0]}, {self.feedback.sequence[1]}")

        # start executing the action
        for i in range(2, goal.order + 1):
            if self.action_server.is_preempt_requested() or rospy.is_shutdown():
                rospy.loginfo(f"{self.action_name}: Preempted")
                self.action_server.set_preempted()
                success = False
                break

            self.feedback.sequence.append(self.feedback.sequence[i - 1] + self.feedback.sequence[i - 2])

            # publish the feedback
            self.action_server.publish_feedback(self.feedback)

            rate.sleep()

        if success:
            self.result.sequence = self.feedback.sequence;
            rospy.loginfo(f"{self.action_name}: Succeeded")
            self.action_server.set_succeeded(self.result)


def main():
    rospy.init_node("fibonacci_server")

    FibonacciAction("fibonacci")
    rospy.loginfo("Node has been started")
    rospy.spin()


if __name__ == "__main__":
    main()
